### PR TITLE
add ability to set image from numpy array

### PR DESCRIPTION
**Pull request by wolfv**

### CODE REVIEW

**Comment by wolfv** [comment]
Hey, so this is an initial version. One thing that I am unsure about is how you do scaling now -- it would be nice to be able to set DPI to something > 1 to get bigger pixels / zoom ...

File: src/widget.ts [hunk 1]
```diff
@@ -113,8 +115,24 @@ class CanvasView extends DOMWidgetView {
   resize_canvas() {
     const size = this.model.get('size');
 
-    this.canvas.setAttribute('width', size[0]);
-    this.canvas.setAttribute('height', size[1]);
+    this.canvas.setAttribute('width', size[0] + 'px');
+    this.canvas.setAttribute('height', size[1] + 'px');
```
Comment (martinRenou) [inline src/widget.ts]: Are you sure about this change? It's not the style attribute, but the `width` and `height` attributes. According to the MDN documentation, it is an integer: https://developer.mozilla.org/en-US/docs/Web/API/HTMLCanvasElement/height

File: src/widget.ts [hunk 2]
```diff
@@ -113,8 +115,24 @@ class CanvasView extends DOMWidgetView {
...
+    if (val !== null && val.shape.length && val.shape[0] > 0)
+    {
+      if (this.canvas.width != val.shape[1] || this.canvas.height != val.shape[0])
+      {
+        this.model.set('size', [val.shape[1], val.shape[0]]);
+      }
+
+      var buffer = new Uint8ClampedArray(val.buffer.buffer);
+      var imd = new ImageData(buffer, val.shape[1], val.shape[0]);
+      var ctx = this.canvas.getContext("2d");
```
Comment (martinRenou) [inline src/widget.ts]: You can access it through `this.ctx`

File: src/widget.ts [hunk 3]
```diff
@@ -113,8 +115,24 @@ class CanvasView extends DOMWidgetView {
...
+    this.canvas.setAttribute('height', size[1] + 'px');
+  }
+
+  image_changed() {
+    var val = this.model.get('image')
+    if (val !== null && val.shape.length && val.shape[0] > 0)
+    {
+      if (this.canvas.width != val.shape[1] || this.canvas.height != val.shape[0])
+      {
+        this.model.set('size', [val.shape[1], val.shape[0]]);
```
Comment (martinRenou) [inline src/widget.ts]: Do we want the size to change when giving more data? Or do we only want to draw on the available pixels?

Comment (martinRenou) [reply src/widget.ts]: I think I would go for the second solution, not having side effects when setting an attribute

**Comment by martinRenou** [comment]
Closing in favor of #8 exposing the `put_image_data` as a method. Thanks for your help on this Wolf!

**Pull request closed without merging**